src/tensorlake/function_executor/allocation_runner/sdk_algorithms.py

```diff
@@ -454,29 +454,67 @@ def validate_and_deserialize_function_call_metadata(
 
 
 def reconstruct_function_call_args(
+    function: Function,
     function_call_metadata: FunctionCallMetadata | ReduceOperationMetadata | None,
     arg_values: Dict[str, Value],
-    function_instance_arg: Any | None,
 ) -> tuple[List[Any], Dict[str, Any]]:
-    """Returns function call args and kwargs reconstructed from arg_values."""
+    """Returns function call args and kwargs reconstructed from arg_values.
+
+    For class methods, inserts the class instance as the first argument (self).
+    """
     if function_call_metadata is None:
         # Application function call created by Server.
         payload_arg: Value = arg_values["application_payload"]
-        args: List[Any] = [payload_arg.object]
-        kwargs: Dict[str, Any] = {}
+        args, kwargs = _reconstruct_application_function_call_args(
+            payload=payload_arg.object,
+            function=function,
+        )
     else:
         # SDK-created function call.
         args, kwargs = _reconstruct_sdk_function_call_args(
             function_call_metadata=function_call_metadata,
             arg_values=arg_values,
         )
 
-    if function_instance_arg is not None:
-        set_self_arg(args, function_instance_arg)
-
     return args, kwargs
 
 
+def _reconstruct_application_function_call_args(
+    payload: Any,
+    function: Function,
+) -> tuple[List[Any], Dict[str, Any]]:
+    """Reconstructs args and kwargs for application function calls.
+
+    Handles three cases:
+    - Zero parameters: returns ([], {})
+    - Single parameter (backward compatible): returns ([payload], {})
+    - Multiple parameters: returns ([], {param_name: value, ...})
+    """
+    signature: inspect.Signature = function_signature(function)
+    params = list(signature.parameters.values())
+
+    # Exclude 'self' parameter for class methods
+    if len(params) > 0 and params[0].name == "self":
+        params = params[1:]
+
+    # Zero parameters case
+    if len(params) == 0:
+        return [], {}
+
+    # Single parameter case (backward compatible)
+    if len(params) == 1:
+        return [payload], {}
+
+    # Multiple parameters case - map payload dict keys to kwargs
+    if not isinstance(payload, dict):
```

Comment:
The payload is coming from user supplied json string. This should be SDKUsageError with a clear explanation on how we got there and what user needs to change in their json payload. We also need test for this as it's going to be a very common error scenario.